### PR TITLE
Iteration over timestamp range (Persistence + Karl). Makefile tweaks.

### PR DESCRIPTION
**Pull request by mzhurovich**

### CODE REVIEW

File: Blocks/Persistence/file.h [hunk 1]
```diff
@@ -312,6 +318,31 @@ class FilePersister {
     }
   }
 
+  std::pair<uint64_t, uint64_t> IndexRangeByTimestampRange(std::chrono::microseconds from,
+                                                           std::chrono::microseconds till) const {
+    std::pair<uint64_t, uint64_t> result{static_cast<uint64_t>(-1), static_cast<uint64_t>(-1)};
```
Comment (dkorolev) [inline Blocks/Persistence/file.h]: Why `{}`, not `()` initialization?:)

File: Blocks/Persistence/file.h [hunk 2]
```diff
@@ -312,6 +318,31 @@ class FilePersister {
...
+    std::lock_guard<std::mutex> lock(file_persister_impl_->mutex);
+    const auto begin_it = std::lower_bound(
```
Comment (dkorolev) [inline Blocks/Persistence/file.h]: Just a thought: Funny you coded it this way. I'd call `std::binary_search` and capture the `from` variable into the lambda :)

File: Blocks/Persistence/file.h (hunk 1 above)
Comment (mzhurovich) [inline Blocks/Persistence/file.h]: Was inspired by your code above :)

File: Blocks/Persistence/file.h [hunk 3]
```diff
@@ -312,6 +318,31 @@ class FilePersister {
...
+        file_persister_impl_->timestamp.begin(),
+        file_persister_impl_->timestamp.end(),
+        from,
+        [](std::chrono::microseconds entry_t, std::chrono::microseconds t) { return entry_t < t; });
+    if (begin_it != file_persister_impl_->timestamp.end()) {
+      result.first = std::distance(file_persister_impl_->timestamp.begin(), begin_it);
+    }
+    if (till.count() != 0) {
```
Comment (dkorolev) [inline Blocks/Persistence/file.h]: Nit: `if (till != std::chrono::microseconds(0))`.

File: Blocks/Persistence/memory.h [hunk 4]
```diff
@@ -154,6 +155,31 @@ class MemoryPersister {
     }
   }
 
+  std::pair<uint64_t, uint64_t> IndexRangeByTimestampRange(std::chrono::microseconds from,
```
Comment (dkorolev) [inline Blocks/Persistence/memory.h]: Duplication, ugh?

File: Blocks/Persistence/test.cc [hunk 5]
```diff
@@ -379,6 +387,7 @@ inline StorableString LargeTestStorableString(int index) {
 template <typename IMPL, int N = 1000>
 void IteratorPerformanceTest(IMPL& impl, bool publish = true) {
   current::time::ResetToZero();
+  using us_t = std::chrono::microseconds;
```
Comment (dkorolev) [inline Blocks/Persistence/test.cc]: :+1:

File: Blocks/Persistence/test.cc [hunk 6]
```diff
@@ -391,21 +400,37 @@ void IteratorPerformanceTest(IMPL& impl, bool publish = true) {
...
+    EXPECT_EQ(100000ll, (*impl.Iterate(us_t(100000), us_t(101000)).begin()).idx_ts.us.count());
+    EXPECT_EQ("0000100 wwwww", (*impl.Iterate(us_t(100000), us_t(101000)).begin()).entry.s);
+  }
 
   // Perftest the creation of a large number of iterators.
   // The test would pass swiftly if the file is being seeked to the right spot,
   // and run forever if every new iteator is scanning the file from the very beginning.
   for (int i = 0; i < N; ++i) {
     const auto cit = impl.Iterate(i, i + 1).begin();
+    EXPECT_EQ(cit, impl.Iterate(us_t(i * 1000), us_t((i + 1) * 1000)).begin());
```
Comment (dkorolev) [inline Blocks/Persistence/test.cc]: Perhaps compare `JSON(cit->idx_ts)`-s instead?

File: Blocks/SS/persister.h [hunk 7]
```diff
@@ -62,12 +62,21 @@ class EntryPersister : public GenericEntryPersister<ENTRY>, public IMPL {
...
   using IterableRange = typename IMPL::IterableRange;
 
   // NOTE: `IMPL::Iterate()` may throw.
   IterableRange Iterate(uint64_t begin, uint64_t end) const { return IMPL::Iterate(begin, end); }
   IterableRange Iterate(uint64_t begin) const { return IMPL::Iterate(begin, static_cast<uint64_t>(-1)); }
+  IterableRange Iterate(std::chrono::microseconds from, std::chrono::microseconds till) const {
+    return IMPL::Iterate(from, till);
+  }
+  IterableRange Iterate(std::chrono::microseconds from) const {
+    return IMPL::Iterate(from, std::chrono::microseconds(0));
```
Comment (dkorolev) [inline Blocks/SS/persister.h]: Any reason the "till the end" convention for `std::chrono::microseconds` is zero, not a cast `-1` ?

File: Blocks/Persistence/memory.h (hunk 4 above)
Comment (mzhurovich) [inline Blocks/Persistence/memory.h]: Yeah, same as for `Iterate()`.

File: Blocks/Persistence/file.h (hunk 3 above)
Comment (mzhurovich) [inline Blocks/Persistence/file.h]: This way it's a bit shorter ;)

File: Blocks/Persistence/memory.h (hunk 4 above)
Comment (dkorolev) [inline Blocks/Persistence/memory.h]: :ok_hand:

File: Karl/karl.h [hunk 8]
```diff
@@ -753,38 +753,35 @@ class GenericKarl final : private KarlStorage<STORAGE_TYPE>,
...
-          // Service is down.
-          // TODO(dkorolev): Graceful shutdown case for `done`.
-          report.currently = current_service_state::down(
-              keepalive.start_time_epoch_microseconds, last_keepalive, e.idx_ts.us, keepalive.uptime);
-        }
-        report.dependencies = keepalive.dependencies;
-        report.runtime = keepalive.runtime;
-        report_for_codename[keepalive.codename] = report;
+    for (const auto& e : keepalives_stream_.InternalExposePersister().Iterate(from, to)) {
+      const claire_status_t& keepalive = e.entry.keepalive;
```
Comment (dkorolev) [inline Karl/karl.h]: Note: GitHub's diff is terrible here. I assume the change is `if (e.idx_ts.us >= from && e.idx_ts.us < to)` -> `Iterate(from, to)`, and it LGTM :)

File: Blocks/SS/persister.h (hunk 7 above)
Comment (mzhurovich) [inline Blocks/SS/persister.h]: Does it really matter? Zero is good enough as any negative number for timestamp, as we use `[from, till)` notation.

File: Blocks/Persistence/test.cc (hunk 6 above)
Comment (mzhurovich) [inline Blocks/Persistence/test.cc]: Hmmm.
```
      // `operator*` relies on the fact each entry will be requested at most once.
      // The range-based for-loop works fine. -- D.K.
```

Comment (dkorolev) [inline Blocks/Persistence/test.cc]: Well, I meant the `const auto& e` one dereferenced down there.

File: Karl/karl.h (hunk 8 above)
Comment (mzhurovich) [inline Karl/karl.h]: Yep :)

**Comment by dkorolev** [comment]
:+1:

**Pull request merged**